## Data collection 
### Files
We have 5 data files in the yelp dataset that we are interested in as shown below

In [2]:
import os
import pandas as pd

dataset_path = 'yelp_dataset'

['.DS_Store', 'yelp_academic_dataset_checkin.json', 'Dataset_User_Agreement.pdf', 'yelp_academic_dataset_tip.json', 'yelp_academic_dataset_review.json', '.ipynb_checkpoints', 'yelp_academic_dataset_business.json', 'yelp_academic_dataset_user.json']


In [3]:
! pip install pyspark
! pip install pyarrow


[notice] A new release of pip is available: 23.0.1 -> 25.2
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.0.1 -> 25.2
[notice] To update, run: pip install --upgrade pip


### checkins file (yelp_academic_dataset_checkin.json)
This file has two columns: 'business_id' and check-in dates column 'date' for that business

In [4]:
checkins_df = pd.read_json(f'{dataset_path}/yelp_academic_dataset_checkin.json', lines=True)
checkins_df.shape

(131930, 2)

In [5]:
checkins_df.head(5)

business_id                                               date
0  ---kPU91CF4Lq2-WlRu9Lw  2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...
1  --0iUa4sNDFiZFrAdIWhZQ  2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...
2  --30_8IhuyMHbSOcNWd6DQ           2013-06-14 23:29:17, 2014-08-13 23:20:22
3  --7PUidqRWpRSpXebiyxTg  2011-02-15 17:12:00, 2011-07-28 02:46:10, 2012...
4  --7jw19RH9JKXgFohspgQw  2014-04-21 20:42:11, 2014-04-28 21:04:46, 2014...

### user tips file (yelp_academic_dataset_tip.json)

In [6]:
tips_df = pd.read_json(f'{dataset_path}/yelp_academic_dataset_tip.json', lines=True)
tips_df.shape

(908915, 5)

In [7]:
tips_df.head()

user_id             business_id  \
0  AGNUgVwnZUey3gcPCJ76iw  3uLgwr0qeCNMjKenHJwPGQ   
1  NBN4MgHP9D3cw--SnauTkA  QoezRbYQncpRqyrLH6Iqjg   
2  -copOvldyKh1qr-vzkDEvw  MYoRNLb5chwjQe3c_k37Gg   
3  FjMQVZjSqY8syIO-53KFKw  hV-bABTK-glh5wj31ps_Jw   
4  ld0AperBXk1h6UbqmM80zw  _uN0OudeJ3Zl_tf6nxg5ww   

                                                text                date  \
0                     Avengers time with the ladies. 2012-05-18 02:17:21   
1  They have lots of good deserts and tasty cuban... 2013-02-05 18:35:10   
2             It's open even when you think it isn't 2013-08-18 00:56:08   
3                          Very decent fried chicken 2017-06-27 23:05:38   
4             Appetizers.. platter special for lunch 2012-10-06 19:43:09   

   compliment_count  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

### user reviews file (yelp_academic_dataset_review.json)
Since the user reviews file is huge (more than 5 GB of data), we can read the file lines as strings to get row count, provided each line represents a row of data.

In [8]:
reviews = f'{dataset_path}/yelp_academic_dataset_review.json'

# Sum up all lines to get row count
with open(reviews, 'r') as file:
    row_count = sum(1 for line in file)

print(f'{row_count} rows in reviews file')

6990280 rows in reviews file


In [9]:
# Load a few rows from user reviews
reviews_iter = pd.read_json(f'{dataset_path}/yelp_academic_dataset_review.json', lines=True, chunksize=1000)
reviews_df = next(reviews_iter)
reviews_df.shape

(1000, 9)

In [10]:
print(f'{reviews_df.shape[1]} columns')

9 columns


In [12]:
reviews_df.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q  7ATYjTIgM3jUlt4UM3IypQ   
2  saUsX_uimxRlCVr67Z4Jig  8g_iMtfSiwikVnbP2etR0A  YjUWPpI6HXG530lwP-fb2A   
3  AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ  kxX2SOes4o-D3ZQBkiMRfA   
4  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   

   stars  useful  funny  cool  \
0      3       0      0     0   
1      5       1      0     1   
2      3       0      0     0   
3      5       1      0     1   
4      4       1      0     1   

                                                text                date  
0  If you decide to eat here, just be aware it is... 2018-07-07 22:09:11  
1  I've taken a lot of spin classes over the year... 2012-01-03 15:28:18  
2  Family diner. Had the buffet. Eclectic assortm... 2014-02-05 20:30:30  
3  Wow!  Yummy, different,  delicious.   Our favo... 2015-01-04 00:01:03  
4  Cute interior and owner (?) gave us tour of up... 2017-01-14 20:54:15

### businesses file (yelp_academic_dataset_business.json)
Our dataset is comprised of 150346 businesses

In [11]:
businesses_df = pd.read_json(f'{dataset_path}/yelp_academic_dataset_business.json', lines=True)
businesses_df.shape          

(150346, 14)

In [12]:
businesses_df.sort_values('review_count', ascending=False).head()

business_id                                name  \
113731  _ab50qdWOk0DdB6XOrBitw                   Acme Oyster House   
112552  ac1AeYqs8Z4_e2X5M3if2A                        Oceana Grill   
91757   GXFMD0Z4jEVZBCsbPf4CTQ  Hattie B’s Hot Chicken - Nashville   
143157  ytynqOUb3hjKeJfRj5Tshw             Reading Terminal Market   
147081  oBNrLz4EDhiscSlbOl8uAw          Ruby Slipper - New Orleans   

                 address          city state postal_code   latitude  \
113731  724 Iberville St   New Orleans    LA       70130  29.954273   
112552      739 Conti St   New Orleans    LA       70130  29.956231   
91757     112 19th Ave S     Nashville    TN       37203  36.151387   
143157      51 N 12th St  Philadelphia    PA       19107  39.953341   
147081   200 Magazine St   New Orleans    LA       70130  29.951025   

        longitude  stars  review_count  is_open  \
113731 -90.068965    4.0          7568        1   
112552 -90.067563    4.0          7400        1   
91757  -86.796603    4.5          6093        1   
143157 -75.158855    4.5          5721        1   
147081 -90.067394    4.5          5193        1   

                                               attributes  \
113731  {'RestaurantsTakeOut': 'True', 'Alcohol': ''fu...   
112552  {'RestaurantsGoodForGroups': 'True', 'Restaura...   
91757   {'RestaurantsGoodForGroups': 'True', 'Business...   
143157  {'RestaurantsGoodForGroups': 'True', 'Restaura...   
147081  {'NoiseLevel': ''loud'', 'Caters': 'False', 'B...   

                                               categories  \
113731  Live/Raw Food, Seafood, Restaurants, Cajun/Creole   
112552  Restaurants, Seafood, Cajun/Creole, Breakfast ...   
91757   American (Traditional), Chicken Shop, Southern...   
143157  Candy Stores, Shopping, Department Stores, Fas...   
147081  Restaurants, American (Traditional), American ...   

                                                    hours  
113731  {'Monday': '11:0-22:0', 'Thursday': '11:0-22:0...  
112552  {'Monday': '8:0-1:0', 'Tuesday': '8:0-1:0', 'W...  
91757   {'Monday': '0:0-0:0', 'Tuesday': '11:0-16:0', ...  
143157  {'Monday': '8:0-18:0', 'Tuesday': '8:0-18:0', ...  
147081  {'Monday': '0:0-0:0', 'Tuesday': '7:30-14:0', ...

#### Finally, how many users are in the Yelp dataset
Users file is also quite large, so we'll just read the line count to get row count

### users file (yelp_academic_dataset_user.json)

In [13]:
users = f'{dataset_path}/yelp_academic_dataset_user.json'

# Sum up all lines to get row count
with open(users, 'r') as file:
    row_count = sum(1 for line in file)

print(f'{row_count} rows in users file')

1987897 rows in users file


In [14]:
# Load a few rows from users file
users_iter = pd.read_json(f'{dataset_path}/yelp_academic_dataset_user.json', lines=True, chunksize=1000)
users_df = next(users_iter)
users_df.shape

(1000, 22)

In [15]:
users_df.columns

Index(['user_id', 'name', 'review_count', 'yelping_since', 'useful', 'funny',
       'cool', 'elite', 'friends', 'fans', 'average_stars', 'compliment_hot',
       'compliment_more', 'compliment_profile', 'compliment_cute',
       'compliment_list', 'compliment_note', 'compliment_plain',
       'compliment_cool', 'compliment_funny', 'compliment_writer',
       'compliment_photos'],
      dtype='object')

In [16]:
users_df.head()

user_id    name  review_count        yelping_since  useful  \
0  qVc8ODYU5SZjKXVBgXdI7w  Walker           585  2007-01-25 16:47:26    7217   
1  j14WgRoU_-2ZE1aw1dXrJg  Daniel          4333  2009-01-25 04:35:42   43091   
2  2WnXYQFK0hXEoTxPtV2zvg   Steph           665  2008-07-25 10:41:00    2086   
3  SZDeASXq7o05mMNLshsdIA    Gwen           224  2005-11-29 04:38:33     512   
4  hA5lMy-EnncsH4JoR-hFGQ   Karen            79  2007-01-05 19:40:59      29   

   funny   cool                                              elite  \
0   1259   5994                                               2007   
1  13066  27281  2009,2010,2011,2012,2013,2014,2015,2016,2017,2...   
2   1010   1003                           2009,2010,2011,2012,2013   
3    330    299                                     2009,2010,2011   
4     15      7                                                      

                                             friends  fans  ...  \
0  NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...   
1  ueRPE0CX75ePGMqOFVj6IQ, 52oH4DrRvzzl8wh5UXyU0A...  3138  ...   
2  LuO3Bn4f3rlhyHIaNfTlnA, j9B4XdHUhDfTKVecyWQgyA...    52  ...   
3  enx1vVPnfdNUdPho6PH_wg, 4wOcvMLtU6a9Lslggq74Vg...    28  ...   
4  PBK4q9KEEBHhFvSXCUirIw, 3FWPpM7KU1gXeOM_ZbYMbA...     1  ...   

   compliment_more  compliment_profile  compliment_cute  compliment_list  \
0               65                  55               56               18   
1              264                 184              157              251   
2               13                  10               17                3   
3                4                   1                6                2   
4                1                   0                0                0   

   compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0              232               844              467               467   
1             1847              7054             3131              3131   
2               66                96              119               119   
3               12                16               26                26   
4                1                 1                0                 0   

   compliment_writer  compliment_photos  
0                239                180  
1               1521               1946  
2                 35                 18  
3                 10                  9  
4                  0                  0  

[5 rows x 22 columns]

### Sample business and reviews

In [17]:
# Select open business with the highest 'review_count'
businesses_df[businesses_df.is_open == 1].sort_values('review_count', ascending=False).head(1)

business_id               name           address  \
113731  _ab50qdWOk0DdB6XOrBitw  Acme Oyster House  724 Iberville St   

               city state postal_code   latitude  longitude  stars  \
113731  New Orleans    LA       70130  29.954273 -90.068965    4.0   

        review_count  is_open  \
113731          7568        1   

                                               attributes  \
113731  {'RestaurantsTakeOut': 'True', 'Alcohol': ''fu...   

                                               categories  \
113731  Live/Raw Food, Seafood, Restaurants, Cajun/Creole   

                                                    hours  
113731  {'Monday': '11:0-22:0', 'Thursday': '11:0-22:0...

### Shape summary (data files)

| Dataset    | Rows      | Columns |
|------------|-----------|---------|
| Users      | 1,987,897 | 22      |
| Reviews    | 6,990,280 | 9       |
| Tips       |   908,915 | 5       |
| Businesses |   150,346 | 14      |
| Checkins   |   131,930 | 2       |